beamer/deploy/util.py

```diff
@@ -0,0 +1,142 @@
+    # constructor_spec is either
+    # 1) a contract name e.g. "Foo", or
+    # 2) a sequence containing the contract name and
+    #    constructor arguments, e.g. ["Foo", "0x1CEE82EEd89Bd5Be5bf2507a92a755dcF1D8e8dc"]
+    if isinstance(constructor_spec, str):
+        name = constructor_spec
+        args: Sequence = ()
+    else:
+        name = constructor_spec[0]
+        args = constructor_spec[1:]
+
+    data = _CONTRACTS[name]
+    log.info("Deploying contract", contract=name)
+    ContractFactory = cast(Contract, web3.eth.contract(abi=data[0], bytecode=data[1]))
+
+    receipt = _transact(ContractFactory.constructor(*args))
+    txhash = encode_hex(receipt.transactionHash)
+
+    address = receipt.contractAddress
+    deployed = cast(DeployedContract, web3.eth.contract(address=address, abi=data[0]))
+    deployed.deployment_block = receipt.blockNumber
+    deployed.deployment_txhash = txhash
+    deployed.deployment_args = list(args)
+    deployed.name = name
+
+    log.info("Deployed contract", contract=name, address=address, txhash=txhash)
+    return deployed
+
+
+def _resolve_constructor_args(resolver: Contract, constructor_spec: str | Sequence) -> tuple:
+    if isinstance(constructor_spec, str):
+        # This is just the contract's name.
+        return (constructor_spec,)
+    return tuple(resolver.address if arg == "${resolver}" else arg for arg in constructor_spec)
+
+
+def deploy_beamer(
+    w3: Web3, chain: config.Chain, resolver: Contract
+) -> tuple[tuple[DeployedContract, ...], tuple[DeployedContract, ...]]:
+    args = _resolve_constructor_args(resolver, chain.l1_messenger)
+    l1_messenger = deploy_contract(resolver.w3, args)
+
+    args = _resolve_constructor_args(resolver, chain.l2_messenger)
+    l2_messenger = deploy_contract(w3, args)
+
+    request_manager = deploy_contract(
+        w3,
+        (
+            "RequestManager",
+            to_wei(chain.request_manager_arguments.claim_stake, "ether"),
+            chain.request_manager_arguments.claim_request_extension,
+            chain.request_manager_arguments.claim_period,
+            chain.request_manager_arguments.challenge_period_extension,
+        ),
+    )
+    fill_manager = deploy_contract(w3, ("FillManager", l2_messenger.address))
+    _transact(fill_manager.functions.setResolver(resolver.address))
+
+    # Authorize call chain
+    _transact(l2_messenger.functions.addCaller(fill_manager.address))
+    _transact(
+        resolver.functions.addCaller(
+            chain.chain_id,
+            l2_messenger.address,
+            l1_messenger.address,
+        )
+    )
+    _transact(
+        resolver.functions.addRequestManager(
+            chain.chain_id, request_manager.address, l1_messenger.address
+        )
+    )
+    _transact(l1_messenger.functions.addCaller(resolver.address))
+    _transact(
+        request_manager.functions.addCaller(
+            resolver.w3.eth.chain_id,
```

Comment:
nitpick: It would be more correct if we'd use `l1_messenger.w3.eth.chain_id` as we effectively set the L1 messenger as the caller in thus the origin chain is where the l1_messenger is (even though the resolver is on the same chain)

Reply:
> nitpick: It would be more correct if we'd use `l1_messenger.w3.eth.chain_id` as we effectively set the L1 messenger as the caller in thus the origin chain is where the l1_messenger is (even though the resolver is on the same chain)

You are absolutely right. Done. :+1: